yt_dlp/extractor/zdf.py

```diff
@@ -410,32 +413,72 @@ def _og_search_title(self, webpage, fatal=False):
+        } for key, thumbnail_info in thumbnails.items() if url_or_none(thumbnail_info.get('url'))]
+
+    def _teaser_to_url_result(self, teaser):
+        return self.url_result(
+            teaser['sharingUrl'], ie=ZDFIE.ie_key(),
+            id=teaser.get('id'), title=teaser.get('titel', ''),
+            thumbnails=self._convert_thumbnails(teaser.get('teaserBild', {})),
+            description=teaser.get('beschreibung'),
+            duration=float_or_none(teaser.get('length')),
+            media_type=teaser.get('currentVideoType') or teaser.get('contentType'),
+            season_number=int_or_none(teaser.get('seasonNumber')),
+            episode_number=int_or_none(teaser.get('episodeNumber')))
+
     def _real_extract(self, url):
         channel_id = self._match_id(url)
 
         webpage = self._download_webpage(url, channel_id)
 
-        matches = re.finditer(
-            rf'''<div\b[^>]*?\sdata-plusbar-id\s*=\s*(["'])(?P<p_id>[\w-]+)\1[^>]*?\sdata-plusbar-url=\1(?P<url>{ZDFIE._VALID_URL})\1''',
-            webpage)
+        main_video = None
+        playlist_videos = []
+
+        document_id = self._extract_document_id(webpage)
+        if document_id is not None:
+            data = self._download_mediathekv2_document(document_id)
+
+            for cluster in data['cluster']:
+                for teaser in cluster['teaser']:
+                    if cluster['type'] == 'teaserContent' and teaser['type'] == 'video':
+                        main_video = main_video or teaser
+                    elif cluster['type'] == 'teaser' and teaser['type'] == 'video':
+                        if teaser['brandId'] != document_id:
+                            # These are unrelated 'You might also like' videos, filter them out
+                            continue
+                        playlist_videos.append(teaser)
 
         if self._downloader.params.get('noplaylist', False):
-            entry = next(
-                (self.url_result(m.group('url'), ie=ZDFIE.ie_key()) for m in matches),
-                None)
-            self.to_screen('Downloading just the main video because of --no-playlist')
-            if entry:
-                return entry
-        else:
-            self.to_screen(f'Downloading playlist {channel_id} - add --no-playlist to download just the main video')
-
-        def check_video(m):
-            v_ref = self._search_regex(
-                r'''(<a\b[^>]*?\shref\s*=[^>]+?\sdata-target-id\s*=\s*(["']){}\2[^>]*>)'''.format(m.group('p_id')),
-                webpage, 'check id', default='')
-            v_ref = extract_attributes(v_ref)
-            return v_ref.get('data-target-video-type') != 'novideo'
-
-        return self.playlist_from_matches(
-            (m.group('url') for m in matches if check_video(m)),
-            channel_id, self._og_search_title(webpage, fatal=False))
+            return self._teaser_to_url_result(main_video) if main_video else None
```

Comment:
Same here: If I can't find a `main_video`, what should I do?

Reply:
addressed above